designs/2020-deprecate-caching/README.md

```diff
@@ -0,0 +1,81 @@
+- Repo: (the repo it will be implemented in, e.g. eslint/espree)
+- Start Date: 2020-10-10
+- RFC PR: (leave this empty, to be filled in later)
+- Authors: fa93hws
+
+# Deprecate caching
+
+## Summary
+Current [eslint caching](https://eslint.org/docs/user-guide/command-line-interface#options) assumes all rules are independent, which is not correct.
+
+## Motivation
+Deprecate `caching` in [options](https://eslint.org/docs/user-guide/command-line-interface#options).
+
+The reason for that is the assumption that all rules are independent is not valid.
+For example, [no-extraneous-dependencies](https://github.com/benmosher/eslint-plugin-import/blob/master/docs/rules/no-extraneous-dependencies.md) has a dependency on `package.json`.
```

Comment:
This is a valid reason, however, resolution seems like using a nuclear bomb to kill flies. How many rules are there between all of the different plugins out there. And what percentage of those rules have external dependency on something? Wouldn't it be better to allow rules to specify a flag that says this rule is not cacheable and modify cache functionality to take that flag into account?

Reply:
I think that would be a alternative solution. But in order to have caching working properly, we may have to set the default value to be `false` to all rules, which means all rules should not use the cache by default. Does that sounds good?

Reply:
I don't think that would be ideal, again, I have no data to support this, but in my estimate, the number of rules that depend on external resources is probably around 2-3% if not less. I don't think it's warranted to completely disable the feature for an edge case. I would set the value to `true` and let the rule creators flip it to false for any rule that depends on external resources.
In this particular case, it's not ESLint's job to ensure that all rules work properly, it's ESLint's job to provide tools for rules to work properly, and leave it to rule creators to use those tools correctly.